internals-js/src/specs/connectSpec.ts

```diff
@@ -0,0 +1,148 @@
+import {DirectiveLocation, GraphQLError} from 'graphql';
+import { CorePurpose, FeatureDefinition, FeatureDefinitions, FeatureUrl, FeatureVersion } from "./coreSpec";
+import {
+  Schema,
+  NonNullType,
+  InputObjectType,
+  InputFieldDefinition,
+  ListType,
+} from '../definitions';
+import { registerKnownFeature } from '../knownCoreFeatures';
+import { createDirectiveSpecification, createScalarTypeSpecification } from '../directiveAndTypeSpecification';
+
+export const connectIdentity = 'https://specs.apollo.dev/connect';
+
+const CONNECT = "connect";
+const SOURCE = "source";
+const URL_PATH_TEMPLATE = "URLPathTemplate";
+const JSON_SELECTION = "JSONSelection";
+const CONNECT_HTTP = "ConnectHTTP";
+const SOURCE_HTTP = "SourceHTTP";
+const HTTP_HEADER_MAPPING = "HTTPHeaderMapping";
+
+export class ConnectSpecDefinition extends FeatureDefinition {
+  constructor(version: FeatureVersion, readonly minimumFederationVersion: FeatureVersion) {
+    super(new FeatureUrl(connectIdentity, CONNECT, version), minimumFederationVersion);
+
+    this.registerDirective(createDirectiveSpecification({
+      name: CONNECT,
+      locations: [DirectiveLocation.FIELD_DEFINITION],
+      repeatable: true,
+      // We "compose" these directives using the  `@join__directive` mechanism,
+      // so they do not need to be composed in the way passing `composes: true`
+      // here implies.
+      composes: false,
+    }));
+
+    this.registerDirective(createDirectiveSpecification({
+      name: SOURCE,
+      locations: [DirectiveLocation.SCHEMA],
+      repeatable: true,
+      composes: false,
+    }));
+
+    this.registerType(createScalarTypeSpecification({ name: URL_PATH_TEMPLATE }));
+    this.registerType(createScalarTypeSpecification({ name: JSON_SELECTION }));
+    this.registerType({ name: CONNECT_HTTP, checkOrAdd: () => [] });
+    this.registerType({ name: SOURCE_HTTP, checkOrAdd: () => [] });
+    this.registerType({ name: HTTP_HEADER_MAPPING, checkOrAdd: () => [] });
+  }
+
+  addElementsToSchema(schema: Schema): GraphQLError[] {
+    /* scalar URLPathTemplate */
+    const URLPathTemplate = this.addScalarType(schema, URL_PATH_TEMPLATE);
+
+    /* scalar JSONSelection */
+    const JSONSelection = this.addScalarType(schema, JSON_SELECTION);
+
+    /*
+      directive @connect(
+        source: String
+        http: ConnectHTTP
+        selection: JSONSelection!
+        entity: Boolean = false
+      ) repeatable on FIELD_DEFINITION
+    */
+    const connect = this.addDirective(schema, CONNECT).addLocations(DirectiveLocation.FIELD_DEFINITION);
+    connect.repeatable = true;
+
+    connect.addArgument(SOURCE, schema.stringType());
+
+    /*
+      input HTTPHeaderMapping {
+        name: String!
+        from: String
+        value: String
+      }
+    */
+    const HTTPHeaderMapping = schema.addType(new InputObjectType(this.typeNameInSchema(schema, HTTP_HEADER_MAPPING)!));
```

Comment:
Ultimately we should add a `createInputObjectTypeSpecification()` function and not override `addElementsToSchema()` to properly get check-or-add functionality, but doing that is more complicated than the other `createXTypeSpecification()` functions, due to some edge cases with references.